### PR TITLE
Fix for supporting ROCm 5.0

### PR DESCRIPTION
**Pull request by takagi**
Rel #6459. Merge #6517 first.

A number of fixes for making CuPy work healthy with ROCm 5.0.

### CODE REVIEW

**Comment by takagi** [comment]
Now ready to review.

File: tests/cupy_tests/core_tests/test_ndarray.py [hunk 1]
```diff
@@ -197,11 +199,14 @@ def test_copy_multi_device_non_contiguous(self, order):
...
         with cuda.Device(1):
-            with self.assertRaises(NotImplementedError):
+            with pytest.raises(NotImplementedError):
                 arr.copy('K')
 
     # See cupy/cupy#5004
     @testing.multi_gpu(2)
+    @pytest.mark.xfail(
+        runtime.is_hip,
+        reason='ROCm may work differently in async D2D copy with streams')
```
Comment (kmaehashi) [inline tests/cupy_tests/core_tests/test_ndarray.py:209]: Question: Is this new in ROCm 5.0?

File: tests/cupy_tests/cuda_tests/test_memory.py [hunk 2]
```diff
@@ -79,7 +79,15 @@ def check(self, device_id):
         if self.specify_device_id:
             kwargs = {'device_id': device_id}
 
-        unowned_mem = memory.UnownedMemory(*args, **kwargs)
+        if cupy.cuda.runtime.is_hip and self.allocator is memory._malloc:
+            # In ROCm, it seems that `hipPointerGetAttributes()`, which is
```
Comment (kmaehashi) [inline tests/cupy_tests/cuda_tests/test_memory.py:83]: Question: is this new in ROCm 5.0?

**Review by kmaehashi — COMMENTED** [review]
Great! LGTM except for a few questions.

File: tests/cupy_tests/core_tests/test_ndarray.py (hunk 1 above)
Comment (takagi) [reply tests/cupy_tests/core_tests/test_ndarray.py:209]: It must be no. I get failed in ROCm 4.3 on my local (with different values compared to that of ROCm 5.0). How about the recent result on Jenkins ROCm tests?

File: tests/cupy_tests/cuda_tests/test_memory.py (hunk 2 above)
Comment (takagi) [reply tests/cupy_tests/cuda_tests/test_memory.py:83]: This test also fails in ROCm 4.3 on my local.

File: tests/cupy_tests/core_tests/test_ndarray.py (hunk 1 above)
Comment (kmaehashi) [reply tests/cupy_tests/core_tests/test_ndarray.py:209]: Ah, Jenkins is not failing because it only has one GPU so this test was not running...

**Review by kmaehashi — APPROVED** [review]

**Comment by kmaehashi** [comment]
@takagi Could you take a look at CUDA failures? :pray:

**Comment by takagi** [comment]
Let me kick the CI.

**Comment by takagi** [comment]
@kmaehashi, can you see the situation of cupy.win.cuda116 that runs so long...?

**Comment by kmaehashi** [comment]
Oops I overlooked a test failure in Jenkins... (can be worked on in a separate PR)


```
11:31:51 FAILED cupyx_tests/scipy_tests/sparse_tests/test_linalg.py::TestLOBPCG::test_maxit_None
...
11:31:51 = 1 failed, 88968 passed, 7328 skipped, 1005 deselected, 4336 xfailed, 160 xpassed, 1733 warnings in 3519.20s (0:58:39) =
```

**Comment by takagi** [comment]
oh... I can't reproduce the failure on my local with ROCm 4.3/5.0 or CUDA 11.5. Can I have the test condition on Jenkins?

**Comment by kmaehashi** [comment]
Maybe the problem is that the test env is still ROCm 5.0.0?
https://github.com/cupy/cupy/blob/master/.pfnci/linux/tests/rocm-5-0.Dockerfile

full log: [consoleText.txt](https://github.com/cupy/cupy/files/8376274/consoleText.txt)

**Comment by takagi** [comment]
Seems the direct cause is that the solver's iteration has finished in fewer times than expected. After I look into how the test works, I've found that CuPy's implementation returns wierd eigenvalues `[9.3633120e+07 1.6104681e+00 8.8367283e-02 1.4951467e-03]` at least in the case of `test_maxit_None` as opposed to SciPy's `[-1.        -1.9999943 -3.000001  -4.000016 ]`. I'm going to see if it reproduces on ROCm 4.3.

**Comment by takagi** [comment]
omg, it works well on ROCm 4.3 with the eigenvalues `[-1.        -2.0000021 -3.0000026 -3.9999971]`. Let me mark the test as XFAIL in ROCm 5.0+ for this release and post an issue.

**Comment by takagi** [comment]
> Maybe the problem is that the test env is still ROCm 5.0.0?

I don't think the patch version 5.0.x is related as far as I read the release notes, but I'll also confirm it.

**Comment by takagi** [comment]
5.0.0 and 5.0.2 make no difference.

**Pull request merged**